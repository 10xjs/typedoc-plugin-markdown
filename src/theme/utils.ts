import * as fs from 'fs';
import * as Handlebars from 'handlebars';
import * as path from 'path';
import { DeclarationReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';

export function compileTemplate(template: string, data: {}) {
  const source = templateToString(template);
  const md: HandlebarsTemplateDelegate = Handlebars.compile(source);
  return new Handlebars.SafeString(md(data));
}

export function safeString(data: string) {
  return new Handlebars.SafeString(data);
}

export function getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

function templateToString(template: string) {
  return fs.readFileSync(path.join(__dirname, `/${template}`)).toString();
}
