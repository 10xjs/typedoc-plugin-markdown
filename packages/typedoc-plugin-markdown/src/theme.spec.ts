import * as fs from 'fs-extra';
import { UrlMapping } from 'typedoc';
import { setup } from '../test/utils';

describe(`theme`, () => {
  function getExpectedUrls(urlMappings: UrlMapping[]) {
    const expectedUrls = [];
    urlMappings.forEach((urlMapping) => {
      expectedUrls.push(urlMapping.url);
      if (urlMapping.model.children) {
        urlMapping.model.children.forEach((reflection) => {
          expectedUrls.push(reflection.url);
        });
      }
    });
    return expectedUrls;
  }

  let app: any;
  let project: any;
  let theme: any;
  beforeAll(() => {
    ({ app, project } = setup());
    theme = app.renderer.theme;
  });

  describe(`getUrls`, () => {
    test(`should getUrls'`, () => {
      const urlMappings = app.renderer.theme.getUrls(project);
      expect(getExpectedUrls(urlMappings)).toMatchSnapshot();
    });
  });

  describe(`output directory`, () => {
    let directoryListingSpy;

    beforeAll(() => {
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'classes', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue(['Unrecognised.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue(['.git', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });

  describe(`output directory`, () => {
    let directoryListingSpy;
    beforeAll(() => {
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'classes', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue(['Unrecognised.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue(['.git', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });
});
