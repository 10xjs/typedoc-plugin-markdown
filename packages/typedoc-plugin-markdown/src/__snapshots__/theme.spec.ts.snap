// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme getUrls should getUrls' 1`] = `
Array [
  "README.md",
  "modules/_categories_.md",
  "modules/_classes_.md",
  "modules/_comments_.md",
  "modules/_functions_.md",
  "modules/_interfaces_.md",
  "modules/_literals_.md",
  "modules/_reflection_.md",
  "modules/_variables_.md",
  "modules/_categories_.md",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.otherinterfacea.md",
  "modules/_categories_.md#categoryvariablea1",
  "modules/_categories_.md#categoryvariablea2",
  "modules/_categories_.md#categoryvariableb1",
  "modules/_categories_.md#othervariablea",
  "modules/_categories_.md#othervariableb",
  "modules/_categories_.md#categoryfunctiona1",
  "modules/_categories_.md#categoryfunctiona2",
  "modules/_categories_.md#categoryfunctiona3",
  "modules/_categories_.md#categoryfunctionb2",
  "modules/_categories_.md#otherfunctiona",
  "modules/_categories_.md#otherfunctionb",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea1.md#value",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.categoryinterfacea2.md#value",
  "interfaces/_categories_.otherinterfacea.md",
  "interfaces/_categories_.otherinterfacea.md#value",
  "modules/_classes_.md",
  "classes/_classes_.greeter.md",
  "modules/_classes_.md#greeter",
  "classes/_classes_.greeter.md",
  "classes/_classes_.greeter.md#constructor",
  "classes/_classes_.greeter.md#greeting",
  "classes/_classes_.greeter.md#greet",
  "modules/_comments_.md",
  "modules/_comments_.md#commentwithdoclinks",
  "modules/_comments_.md#commentswithfencedblock",
  "modules/_comments_.md#commentswithincludes",
  "modules/_comments_.md#commentswithsymbollinks",
  "modules/_comments_.md#commentswithtags",
  "modules/_comments_.md#commentsinreturn",
  "modules/_functions_.md",
  "modules/_functions_.md#buildname",
  "modules/_functions_.md#functionwithdefaults",
  "modules/_functions_.md#functionwithoptionalvalue",
  "modules/_functions_.md#functionwithparameters",
  "modules/_functions_.md#functionwithrest",
  "modules/_functions_.md#modulefunction",
  "modules/_functions_.md#multiplesignatures",
  "modules/_functions_.md#restuntiontypes",
  "modules/_functions_.md#variablefunction",
  "modules/_interfaces_.md",
  "modules/_literals_.md",
  "modules/_literals_.md#easing",
  "modules/_literals_.md#result",
  "modules/_literals_.md#typeliteral",
  "modules/_literals_.md#typeliteral2",
  "modules/_literals_.md#rolldice",
  "modules/_literals_.md#objectliteral",
  "modules/_literals_.md#objectliteral2",
  "modules/_reflection_.md",
  "classes/_reflection_.childclassa.md",
  "classes/_reflection_.childclassb.md",
  "classes/_reflection_.parentclass.md",
  "interfaces/_reflection_.callableinterface.md",
  "interfaces/_reflection_.indexableinterface.md",
  "interfaces/_reflection_.typeparamsinterface.md",
  "classes/_reflection_.childclassa.md",
  "classes/_reflection_.childclassa.md#prop",
  "classes/_reflection_.childclassb.md",
  "classes/_reflection_.childclassb.md#prop",
  "classes/_reflection_.parentclass.md",
  "classes/_reflection_.parentclass.md#prop",
  "interfaces/_reflection_.callableinterface.md",
  "interfaces/_reflection_.indexableinterface.md",
  "interfaces/_reflection_.typeparamsinterface.md",
  "interfaces/_reflection_.typeparamsinterface.md#prop",
  "modules/_variables_.md",
  "modules/_variables_.md#color",
  "modules/_variables_.md#decimal",
  "modules/_variables_.md#isdone",
  "modules/_variables_.md#list",
  "modules/_variables_.md#n",
  "modules/_variables_.md#notsure",
  "modules/_variables_.md#u",
  "modules/_variables_.md#x",
]
`;
